File: 백준/Gold/32358. 근성아 일하자/근성아 일하자.py
# https://www.acmicpc.net/problem/32358
import sys
input = sys.stdin.readline
INF = int(1e9)

wastes = [-INF, INF]
now, total = 0, 0
for _ in range(int(input())):
    q, *x = map(int, input().split())
    if q == 1:
        wastes.append(x[0])
        continue

    wastes.sort()
    left, right = 0, 0
    for i in range(len(wastes)-1):
        if wastes[i] <= now <= wastes[i+1]:
            left, right = i, i+1

    for _ in range(len(wastes)-2):
        if now - wastes[left] <= wastes[right] - now:
            total += now - wastes[left]
            now = wastes[left]
            left -= 1
        else:
            total += wastes[right] - now
            now = wastes[right]
            right += 1

    wastes = [-INF, INF]

print(total)

File: 백준/Gold/32358. 근성아 일하자/근성아 일하자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        int INF = (int)1e9;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int tc = Integer.parseInt(br.readLine());
        List<Integer> numbers = new ArrayList<>(List.of(-INF, INF));
        int now = 0;
        long total = 0L;

        while (tc-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int q = Integer.parseInt(st.nextToken());

            if (q == 1) {
                numbers.add(Integer.parseInt(st.nextToken()));
                continue;
            }

            Collections.sort(numbers);
            int left = 0, right = 0;
            for (int i = 0; i < numbers.size()-1; i++) {
                if (numbers.get(i) <= now && now <= numbers.get(i+1)) {
                    left = i;
                    right = i+1;
                    break;
                }
            }

            for (int unused = 0; unused < numbers.size()-2; unused++) {
                if (now - numbers.get(left) <= numbers.get(right) - now) {
                    total += now - numbers.get(left);
                    now = numbers.get(left--);
                } else {
                    total += numbers.get(right) - now;
                    now = numbers.get(right++);
                }
            }

            numbers = new ArrayList<>(List.of(-INF, INF));
        }

        System.out.println(total);
    }
}